### ISSUE TITLE
Question about using distributed loss 

### ISSUE THREAD

**Issue by liangyuandg** [KevinMusgrave/pytorch-metric-learning#302]
Hi,

I am working with torch.DistributedDataParallel. I am following your /pytorch-metric-learning/examples/notebooks/DistributedTripletMarginLossMNIST.ipynb. I noticed that all the gpus have the same loss value in each epoch. I am trying to understand why. 

According to torch.DistributedDataParallel, each gpu will get its replica of data, and calculate the loss with that part of data. The losses among different gpus are not reduced (the gradients do). Thus, different gpus should have different loss values. My question is, how is the loss calculated here? Is it an average of all gpus' losses? 

Thanks.

**Comment by KevinMusgrave** [comment]
It's possible there is a bug! I will look into it.

**Comment by KevinMusgrave** [comment]
I finally got a chance to check if this is actually working or not. Long story short, each GPU has the same loss value, but different gradients. You can confirm this if you **don't** wrap your model with DistributedDataParallel, and then put this print statement after loss.backward():

```python
print(f"rank {rank} gradients", [p.grad for p in model.parameters()])
```

Here's why the loss values are the same but the gradients are different:

- gpu1 has emb1
- gpu2 has emb2

To compute the loss, gpu1 will collect emb2 from gpu2. However, emb2 comes without gradients.

- gpu1 computes the loss using [emb1, emb2], but only emb1 has gradients
- gpu2 computes the loss using [emb1, emb2], but only emb2 has gradients
